support-frontend/assets/components/asyncronously/asyncronously.ts

```diff
@@ -1,47 +1,38 @@
-import type { ComponentType, Node } from 'react';
-import { Component } from 'react';
+import type { ComponentType } from 'preact';
+import { useEffect, useRef, useState } from 'preact/hooks';
 
-type PropTypes = {
+type PropTypes<T> = {
 	loader: Promise<{
-		default: ComponentType<any>;
+		default: ComponentType<T>;
 	}>;
-	loading: Node;
-	children: (arg0: ComponentType<any>) => Node;
-};
-type State = {
-	loaded: boolean;
+	loading: JSX.Element;
+	render: (MktConsent: ComponentType<T>) => JSX.Element;
 };
 
-class Asyncronously extends Component<PropTypes, State> {
-	static defaultProps = {
-		loading: null,
-	};
-	state = {
-		loaded: false,
-	};
-
-	componentDidMount() {
-		const { loader } = this.props;
-		loader.then((imported) => {
-			this.component = imported.default;
-			this.setState({
-				loaded: true,
-			});
-		});
-	}
+function Asyncronously<T>(props: PropTypes<T>): JSX.Element | null {
+	const [loaded, setLoaded] = useState(false);
 
-	component = null;
+	const componentRef = useRef<ComponentType<T>>();
```

Comment:
Replacing class based instance variable `component = null` with a ref

Reply:
does this need to be a ref? could we use `useState` instead?

Reply:
The reason for this is because `imported` _is_ already a function. When you pass a function to `setState` react will invoke the function and then set the returned value as the new state.

Reply:
Had taken a similar approach to your suggestion at first but came across some mention online that storing a component in state is bad practice, though I suppose in this case using a ref is only marginally better

Reply:
Ideally we could have used React lazy and suspense here instead but it's still experimental with preact. 

From what I can tell, storing a component in state is mainly discouraged because it's an 'anti pattern'. Though I suppose both approaches would be to some degree. Do you think one approach is more beneficial than the other?

Reply:
Yeah it does feel like we're rolling our own suspense mode here!

Did anywhere mention why it's an 'anti pattern'? I can see lots of people saying not to do this, but none of their reasons apply to our use case (showing something else whilst we async load in the actual component). 

I'd be tempted to use `useState` over `useRef`. Mainly because that's how we do it for the [epics & banners in DCR](https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering%2Fsrc%2Fweb%2Fcomponents%2FStickyBottomBanner%2FReaderRevenueBanner.tsx#L289).

It would also allow to drop the `loaded` state too. I think that's nice because it's generally good to structure your state such that there are no invalid states. Just looking at the state (i.e ignoring the logic in the component), it's possible to have `loaded = false` but have `componentRef.current` set, which is an invalid config. 

If instead we just have some state which is of type `ComponentType | undefined`, it's now no longer possible to have this invalid config. 

This is less explicit though. We're now implicitly treating `undefined` as our loading state. If you want to be more explicit, tagged unions are great for this. You could do something like:

```ts
type Status = { type: 'loading' } | { 'type': 'loaded', component: ComponentType }
```

Reply:
Anti pattern wise this is one of the better [explanations I came across](https://stackoverflow.com/a/62913227), essentially if what we return from the render function has side effects then it may produce unexpected results. Having tried both approaches they both seemed to do what we want so if you feel strongly about the changeover that's all good with me. 

Yeah the logic and the state is quite closely coupled in this component at the moment. It did just seem like there was a lot of push back out there for not storing components in state but not much on alternate approaches when you're not using lazy and suspense.

I reckon as it's a fairly small/readable component we don't think we need to be super explicit about the state?